reader-v2.go

```diff
@@ -0,0 +1,174 @@
+// Copyright (C) 2018 Minio Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sio
+
+import (
+	"io"
+)
+
+type encReaderV20 struct {
+	authEncV20
+	src io.Reader
+
+	buffer   packageV20
+	offset   int
+	lastByte byte
+
+	firstRead bool
+}
+
+func encryptReaderV20(src io.Reader, config *Config) (*encReaderV20, error) {
+	ae, err := newAuthEncV20(config)
+	if err != nil {
+		return nil, err
+	}
+	return &encReaderV20{
+		authEncV20: ae,
+		src:        src,
+		buffer:     make(packageV20, maxPackageSize),
+		firstRead:  true,
+	}, nil
+}
+
+func (r *encReaderV20) Read(p []byte) (n int, err error) {
+	if r.firstRead {
+		r.firstRead = false
+		_, err = io.ReadFull(r.src, r.buffer[headerSize:headerSize+1])
+		if err != nil && err != io.EOF { // ErrUnexpectedEOF cannot happen b/c we read just one single byte
+			return 0, err
+		}
+		if err == io.EOF {
+			r.finalized = true
+			return 0, io.EOF
+		}
+		r.lastByte = r.buffer[headerSize]
+	}
+
+	if r.offset > 0 { // write the buffered package to p
+		remaining := r.buffer.Length() - r.offset
+		if len(p) < remaining {
+			r.offset += copy(p, r.buffer[r.offset:r.offset+len(p)])
+			return len(p), nil
+		}
+		n = copy(p, r.buffer[r.offset:r.offset+remaining])
+		p = p[remaining:]
+		r.offset = 0
+	}
+	if r.finalized {
+		return n, io.EOF
+	}
+	for len(p) >= maxPackageSize {
+		r.buffer[headerSize] = r.lastByte
+		nn, err := io.ReadFull(r.src, r.buffer[headerSize+1:headerSize+1+maxPayloadSize]) // try to read the max. payload
+		if err != nil && err != io.EOF && err != io.ErrUnexpectedEOF {
+			return n, err // failed to read from src
+		}
+		if err == io.EOF || err == io.ErrUnexpectedEOF { // read less than 64KB -> final package
+			r.SealFinal(p, r.buffer[headerSize:headerSize+1+nn])
+			return n + headerSize + tagSize + 1 + nn, io.EOF
+		}
+		r.lastByte = r.buffer[headerSize+maxPayloadSize] // save last read byte for the next package
+		r.Seal(p, r.buffer[headerSize:headerSize+maxPayloadSize])
+		p = p[maxPackageSize:]
+		n += maxPackageSize
+	}
+	if len(p) > 0 {
+		r.buffer[headerSize] = r.lastByte
+		nn, err := io.ReadFull(r.src, r.buffer[headerSize+1:headerSize+1+maxPayloadSize]) // try to read the max. payload
```

Comment:
You could simply do perhaps style pattern?

```
if nn > 0 {
}
```

Reply:
Meaning?
There are the following cases according to `io.ReadFull`:
 - `n == 0` **iff** `err == io.EOF` (assuming the passed buffer > 0)
 - underlying io.Reader returns non-EOF error -> `n > 0` and `err != nil && err != io.ErrUnexpectedEOF`
 - underlying io.Reader returns EOF after reading less than requested bytes -> `n > 0` and `err == io.ErrUnexpectedEOF`

So `n > 0` is ambiguous - it can be either the second case in which we are supposed to error out, or the third case in which we are supposed to proceed.

Reply:
yes but io.ErrUnexpectedEOF is not an error @aead using `io.ReadFull` has this drawback.

Reply:
Yes, but I still have to distinguish whether the `maxPayloadSize > n > 0` and the underlying reader returned EOF - which means I get `io.ErrUnexpectedEOF` or whether the underlying reader was able to read  `maxPayloadSize > n > 0` and than an error (e.g. network error) occurred.
In the first case I have to proceed, seal the data and return io.EOF - in the later case I have to error out. How should I distinguish both cases without checking the returned error from `io.ReadFull`?